File: testbed/test/cruise/associations/ImmutableTest.java
package cruise.associations;

import org.junit.*;
import java.lang.reflect.*;

public class ImmutableTest
{
  @Test
  public void mutableClassHasSettersRemoversAndAddMethods()
  {
    MentorA mentor = new MentorA();
    Assert.assertTrue(objectClassHasSettersAddersOrRemovers(mentor));
  }

  @Test
  public void unidirectionalManyAssociation()
  {
    Otherclass other1 = new Otherclass("otherClass");
    Otherclass other2 = new Otherclass("otherClass");

    ClassManyAssociation many = new ClassManyAssociation();
    Assert.assertFalse(many.hasOtherclasses());
    many = new ClassManyAssociation(other1);
    Assert.assertTrue(many.hasOtherclasses());
    many.delete();
    Assert.assertTrue(many.hasOtherclasses());

    many = new ClassManyAssociation(other1, other2);
    Assert.assertTrue(many.hasOtherclasses());
    Assert.assertTrue(many.indexOfOtherclass(other1) >= 0);
    Assert.assertTrue(many.indexOfOtherclass(other2) >= 0);
    Assert.assertFalse(objectClassHasSettersAddersOrRemovers(many));
  }

  @Test
  public void unidirectionalMToNAssociation()
  {
    Otherclass other1 = new Otherclass("other");
    Otherclass other2 = new Otherclass("other");
    Otherclass other3 = new Otherclass("other");
    Otherclass other4 = new Otherclass("other");

    ClassMToNAssociation mToN;
    try {
      mToN = new ClassMToNAssociation();
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}

    try {
      mToN = new ClassMToNAssociation(other1);
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}

    mToN = new ClassMToNAssociation(other1, other2);
    Assert.assertTrue(mToN.hasOtherclasses());
    Assert.assertTrue(mToN.indexOfOtherclass(other1) >= 0);
    Assert.assertTrue(mToN.indexOfOtherclass(other2) >= 0);
    mToN.delete();
    Assert.assertTrue(mToN.hasOtherclasses());

    mToN = new ClassMToNAssociation(other1, other2, other3);
    Assert.assertTrue(mToN.hasOtherclasses());

    try {
      mToN = new ClassMToNAssociation(other1, other2, other3, other4);
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}
    
    Assert.assertFalse(objectClassHasSettersAddersOrRemovers(mToN));
  }

  @Test
  public void unidirectionalMStarAssociation()
  {
    Otherclass other1 = new Otherclass("other");
    Otherclass other2 = new Otherclass("other");
    Otherclass other3 = new Otherclass("other");

    ClassMStarAssociation mStar;
    try {
      mStar = new ClassMStarAssociation();
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}

    try {
      mStar = new ClassMStarAssociation(other1);
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}

    mStar = new ClassMStarAssociation(other1, other2);
    Assert.assertTrue(mStar.hasOtherclasses());
    Assert.assertTrue(mStar.indexOfOtherclass(other1) >= 0);
    Assert.assertTrue(mStar.indexOfOtherclass(other2) >= 0);
    mStar.delete();
    Assert.assertTrue(mStar.hasOtherclasses());

    mStar = new ClassMStarAssociation(other1, other2, other3);
    Assert.assertTrue(mStar.hasOtherclasses());
    
    Assert.assertFalse(objectClassHasSettersAddersOrRemovers(mStar));
  }

  @Test
  public void unidirectionalNAssociation()
  {
    Otherclass other1 = new Otherclass("other");
    Otherclass other2 = new Otherclass("other");
    Otherclass other3 = new Otherclass("other");

    ClassNAssociation n;
    try {
      n = new ClassNAssociation();
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}

    try {
      n = new ClassNAssociation(other1);
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}

    n = new ClassNAssociation(other1, other2);
    Assert.assertTrue(n.hasOtherclasses());
    Assert.assertTrue(n.indexOfOtherclass(other1) >= 0);
    Assert.assertTrue(n.indexOfOtherclass(other2) >= 0);
    n.delete();
    Assert.assertTrue(n.hasOtherclasses());

    try {
      n = new ClassNAssociation(other1, other2, other3);
    } catch (RuntimeException ex) {}
    
    Assert.assertFalse(objectClassHasSettersAddersOrRemovers(n));
  }

  @Test
  public void unidirectionalOneAssociation()
  {
    Otherclass other = new Otherclass("otherClass");

    ClassOneAssociation one = new ClassOneAssociation(other);
    Assert.assertTrue(one.getOtherclass() == other);
    one.delete();
    Assert.assertTrue(one.getOtherclass() == other);

    try {
      one = new ClassOneAssociation(null);
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}
    
    Assert.assertFalse(objectClassHasSettersAddersOrRemovers(one));  
  }

  @Test
  public void unidirectionalOptionalNAssociation()
  {
    Otherclass other1 = new Otherclass("other");
    Otherclass other2 = new Otherclass("other");
    Otherclass other3 = new Otherclass("other");

    ClassOptionalNAssociation optN = new ClassOptionalNAssociation();
    Assert.assertFalse(optN.hasOtherclasses());
    
    optN = new ClassOptionalNAssociation(other1);
    Assert.assertTrue(optN.hasOtherclasses());
    Assert.assertTrue(optN.indexOfOtherclass(other1) >= 0);

    optN = new ClassOptionalNAssociation(other1, other2);
    Assert.assertTrue(optN.hasOtherclasses());
    Assert.assertTrue(optN.indexOfOtherclass(other1) >= 0);
    Assert.assertTrue(optN.indexOfOtherclass(other2) >= 0);
    optN.delete();
    Assert.assertTrue(optN.hasOtherclasses());

    try {
      optN = new ClassOptionalNAssociation(other1, other2, other3);
      Assert.fail("Constructor should have thrown RuntimeException");
    } catch (RuntimeException ex) {}

    Assert.assertFalse(objectClassHasSettersAddersOrRemovers(optN));   
  }

  @Test
  public void unidirectionalOptionalOneAssociation()
  {
    Otherclass other = new Otherclass("otherClass");

    ClassOptionalOneAssociation optOne = new ClassOptionalOneAssociation(other);
    Assert.assertTrue(optOne.getOtherclass() == other);
    optOne.delete();
    Assert.assertTrue(optOne.getOtherclass() == other);

    optOne = new ClassOptionalOneAssociation(null);
    Assert.assertTrue(optOne.getOtherclass() == null);
    
    Assert.assertFalse(objectClassHasSettersAddersOrRemovers(optOne));    
  }


  private boolean objectClassHasSettersAddersOrRemovers(Object obj)
  {
    boolean hasMethods = false;

    Class<? extends Object> clazz = obj.getClass();
    Method[] methods = clazz.getMethods();
    String setterNameRegex = "(set|add|remove)[A-Z]+[a-zA-Z]*";

    for (Method m : methods)
    {
      if (m.getName().matches(setterNameRegex))
      {
        hasMethods = true;
        break;
      }
    }
    return hasMethods;
  }

}

File: testbed_php/test/associations/ImmutableTest.php
<?php

class AssociationImmutableTest extends UnitTestCase
{

  public function test_mutableClassHasSettersRemoversAndAddMethodsAssociation()
  {
    $mentor = new MentorA();
    $this->assertTrue($this->objectClassHasSettersAddersOrRemoversAssociation($mentor));
  }

  
  public function test_unidirectionalManyAssociation()
  {
    $other1 = new Otherclass("otherClass1");
    $other2 = new Otherclass("otherClass2");

    $many = new ClassManyAssociation(array());
    $this->assertFalse($many->hasOtherclasses());
    $many = new ClassManyAssociation(array($other1));
    $this->assertTrue($many->hasOtherclasses());
    $many = new ClassManyAssociation(array($other1, $other2));
    $many->delete();
    $this->assertTrue($many->hasOtherclasses());

    $this->assertTrue($many->hasOtherclasses());
    $this->assertTrue($many->indexOfOtherclass($other1) >= 0);
    $this->assertTrue($many->indexOfOtherclass($other2) >= 0);
    $this->assertFalse($this->objectClassHasSettersAddersOrRemoversAssociation($many));
  }

  
  public function test_unidirectionalMToNAssociation()
  {
    $other1 = new Otherclass("other1");
    $other2 = new Otherclass("other2");
    $other3 = new Otherclass("other3");
    $other4 = new Otherclass("other4");

    try {
      $mToN = new ClassMToNAssociation(array());
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    try {
      $mToN = new ClassMToNAssociation(array($other1));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    $mToN = new ClassMToNAssociation(array($other1,$other2));
    $this->assertTrue($mToN->hasOtherclasses());
    $this->assertTrue($mToN->indexOfOtherclass($other1) >= 0);
    $this->assertTrue($mToN->indexOfOtherclass($other2) >= 0);

    $mToN = new ClassMToNAssociation(array($other1,$other2,$other3));
    $this->assertTrue($mToN->hasOtherclasses());
    $mToN->delete();
    $this->assertTrue($mToN->hasOtherclasses());

    try {
      $mToN = new ClassMToNAssociation(array($other1,$other2,$other3,$other4));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemoversAssociation($mToN));
  }

  
  public function test_unidirectionalMStarAssociation()
  {
    $other1 = new Otherclass("other1");
    $other2 = new Otherclass("other2");
    $other3 = new Otherclass("other3");

    try {
      $mStar = new ClassMStarAssociation(array());
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    try {
      $mStar = new ClassMStarAssociation(array($other1));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    $mStar = new ClassMStarAssociation(array($other1,$other2));
    $this->assertTrue($mStar->hasOtherclasses());
    $this->assertTrue($mStar->indexOfOtherclass($other1) >= 0);
    $this->assertTrue($mStar->indexOfOtherclass($other2) >= 0);

    $mStar = new ClassMStarAssociation(array($other1,$other2,$other3));
    $this->assertTrue($mStar->hasOtherclasses());
    $mStar->delete();
    $this->assertTrue($mStar->hasOtherclasses());
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemoversAssociation($mStar));
  }

  
  public function test_unidirectionalNAssociation()
  {
    $other1 = new Otherclass("other1");
    $other2 = new Otherclass("other2");
    $other3 = new Otherclass("other3");

    try {
      $n = new ClassNAssociation(array());
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    try {
      $n = new ClassNAssociation(array($other1));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    $n = new ClassNAssociation(array($other1,$other2));
    $this->assertTrue($n->hasOtherclasses());
    $this->assertTrue($n->indexOfOtherclass($other1) >= 0);
    $this->assertTrue($n->indexOfOtherclass($other2) >= 0);
    $n->delete();
    $this->assertTrue($n->hasOtherclasses());

    try {
      $n = new ClassNAssociation(array($other1,$other2,$other3));
    } catch (Exception $ex) {}
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemoversAssociation($n));
  }

  
  public function test_unidirectionalOneAssociation()
  {
    $other = new Otherclass("otherClass");

    $one = new ClassOneAssociation($other);
    $this->assertTrue($one->getOtherclass() == $other);
    $one->delete();
    $this->assertTrue($one->getOtherclass() == $other);

    try {
      $one = new ClassOneAssociation(null);
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemoversAssociation($one));  
  }

  
  public function test_unidirectionalOptionalNAssociation()
  {
    $other1 = new Otherclass("other1");
    $other2 = new Otherclass("other2");
    $other3 = new Otherclass("other3");

    $optN = new ClassOptionalNAssociation(array());
    $this->assertFalse($optN->hasOtherclasses());
    
    $optN = new ClassOptionalNAssociation(array($other1));
    $this->assertTrue($optN->hasOtherclasses());
    $this->assertTrue($optN->indexOfOtherclass($other1) >= 0);

    $optN = new ClassOptionalNAssociation(array($other1,$other2));
    $this->assertTrue($optN->hasOtherclasses());
    $this->assertTrue($optN->indexOfOtherclass($other1) >= 0);
    $this->assertTrue($optN->indexOfOtherclass($other2) >= 0);
    $optN->delete();
    $this->assertTrue($optN->hasOtherclasses());

    try {
      $optN = new ClassOptionalNAssociation(array($other1,$other2,$other3));
      $this->fail("Constructor should have thrown RuntimeException");
    } catch (Exception $ex) {}

    $this->assertFalse($this->objectClassHasSettersAddersOrRemoversAssociation($optN));   
  }

  
  public function test_unidirectionalOptionalOneAssociation()
  {
    $other = new Otherclass("otherClass");

    $optOne = new ClassOptionalOneAssociation($other);
    $this->assertTrue($optOne->getOtherclass() == $other);
    $optOne->delete();
    $this->assertTrue($optOne->getOtherclass() == $other);

    $optOne = new ClassOptionalOneAssociation(null);
    $this->assertTrue($optOne->getOtherclass() == null);
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemoversAssociation($optOne));    
  }
  

  private function objectClassHasSettersAddersOrRemoversAssociation($obj)
  {
    $hasMethods = false;

    $clazz = new ReflectionClass(get_class($obj));
    $methods = $clazz->getMethods();

    foreach ($methods as $m)
    {
      if ((preg_match('/^(set|add|remove)[A-Z]+[a-zA-Z]*$/', $m->getName()) > 0) && $m->isPublic())
      {
        $hasMethods = true;
        break;
      }
    }
    return $hasMethods;
  }
}

?>
